File: CTCI/Interview Questions/C++/4.2_minimal_tree.cpp
// Given a sorted (increasing order) array with unique integer elements, write an algorithm to create a binary search tree with minimal height.
// Page 75

#include <iostream>
#include <vector>
#include <algorithm>
#include <ctime>
using namespace std;

struct Node{
    int data;
    Node *left = nullptr;
    Node *right = nullptr;
    Node(int newdata){
        data = newdata;
    }
};

class BinarySearchTree{
private:
    Node * root = nullptr;
public:
    BinarySearchTree(vector <int> &values){
        construct_bst(values, 0, values.size() - 1);
    }
    // Solution
    // Kinda like binary search. Pretty simple solution.
    // Time Complexity: O(n)
    Node * construct_bst(vector <int> &elements, int start, int end){
        if(end < start) return nullptr;
        int mid = (start + end) / 2;
        Node * new_node = new Node(elements[mid]);
        if(root == nullptr) root = new_node;
        new_node -> left = construct_bst(elements, start, mid - 1);
        new_node -> right = construct_bst(elements, mid + 1, end);
        return new_node;
    }
    Node * get_root(){
        return root;
    }
};

void inorder(Node * node){
    if(node == NULL) return;
    inorder(node -> left);
    cout << node -> data << " ";
    inorder(node -> right);
}

vector <int> gen_array(int n, int upper_bound=100){
    if(n >= upper_bound) return vector <int> {};
    vector <int> elements;
    while(elements.size() < n){
        int rand_num = rand() % 100;
        if(find(elements.begin(), elements.end(), rand_num) == elements.end())
            elements.push_back(rand_num);
    }
    sort(elements.begin(), elements.end());
    return elements;
}

int main(void){
    srand(time(NULL));
    int num_elements = 10;
    vector <int> arr = gen_array(num_elements);
    BinarySearchTree * bst = new BinarySearchTree(arr);
    inorder(bst -> get_root());
}

File: CTCI/Interview Questions/Python/4.2_minimal_tree.py
# Given a sorted (increasing order) array with unique integer elements, write an algorithm to create a binary search tree with minimal height.
# Page 75

import random

class Node:
    def __init__(self, value):
        self.val = value
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self, values):
        self.root = None
        self.construct_bst(values, 0, len(values) - 1)

    # Solution
    # Kinda like binary search. Pretty simple solution.
    # Time Complexity: O(n)
    def construct_bst(self, values, start, end):
        if end < start:
            return None
        mid = (start + end) // 2
        node = Node(values[mid])
        if self.root is None:
            self.root = node
        node.left = self.construct_bst(values, start, mid - 1)
        node.right = self.construct_bst(values, mid + 1, end)
        return node

    def inorder(self, node):
        if node is None:
            return
        self.inorder(node.left)
        print(node.val, end=' ')
        self.inorder(node.right)

def gen_array(n, upper_bound=100):
    """Generates and returns an array that is sorted without any duplicates."""
    if n >= upper_bound:
        raise Exception('Upper bound is too high! Either increase it or increase the number of elements in the array')
    arr = []
    while len(arr) <= n:
        rand_num = random.randint(1, upper_bound)
        if rand_num not in arr:
            arr.append(rand_num)
    return sorted(arr)

arr = gen_array(10)
bst = BinarySearchTree(arr)
bst.inorder(bst.root)
